File: #116_Populating_Next_Right_Pointers_in_Each_Node/Solution.java
import java.util.*;

class TreeLinkNode {
	int val;
	TreeLinkNode left, right, next;
	TreeLinkNode(int x) { val = x; }
}

public class Solution {
	public void connect(TreeLinkNode root) {
		if (root == null) return;
		TreeLinkNode node = root;
		TreeLinkNode cur = null;
		while (node != null) {
			cur = node;
			while (cur != null) {
				if (cur.left == null) return;
				cur.left.next = cur.right;
				if (cur.next != null) cur.right.next = cur.next.left;
				cur = cur.next;
			}
			node = node.left;
		}
	}
}

File: #116_Populating_Next_Right_Pointers_in_Each_Node/Solution.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
	# @param root, a tree link node
	# @return nothing
	def connect(self, root):
		if not root:
			return
		cur = root
		startNode = root
		while True:
			if cur is startNode:
				startNode = cur.left
				if not startNode:
					return
			cur.left.next = cur.right
			tmp = cur.right
			cur = cur.next
			if cur:
				tmp.next = cur.left
			else:
				cur = startNode